[DR-2925] Stabilize DRS tests impacted by slow IAM propagation

okotsopoulos opened the pull request:
https://broadworkbench.atlassian.net/browse/DR-2925

Our DRS integration tests have been experiencing frequent failures due to Google-side delays in IAM propagation.   [Doug's finding](https://broadinstitute.slack.com/archives/C53JYBV9A/p1675700250147059?thread_ts=1675673871.679629&cid=C53JYBV9A) has been that our traditional approach of polling for access has not been working and in fact does harm, poisoning Google's cache.

**Original Attempt: Wait Before Checking**

I first tried to follow Doug's recommendations which have helped stabilize tests elsewhere in Terra:

- Commit 1: We now wait 3 minutes before checking for BQ access in test assertions driving current flakiness in DRS tests to avoid poisoning Google's cache.
- Commit 2: After our initial 3 minute wait, don't bother to poll for access.  If the resource is inaccessible, then fail the test.

Unfortunately we still experienced test instability with these changes.  It's possible that the permission we require -- `bigquery.jobs.create` -- takes longer to propagate than the ones required in Terra tests.

**Final Approach: Cut Out Direct BQ Access**

After discussion with the team, we also now leverage the snapshot preview endpoint to fetch a snapshot's DRS IDs rather than querying BigQuery directly.  Because TDR proxies the request to BigQuery, this workaround provides a tangible stability boost.

We've also disabled the accessibility check previously used in DRS test set up: we know that IAM propagation is unreliable, so the assertion's value does not outweigh the reduced stability and increased test runtime.

**Impact on Runtime**

Here are some example runtimes of the DrsTest suite over the course of this work.

Prior to this PR (polled for access):
- **3h 12m** runtime with flaky and failing tests ([Gradle](https://scans.gradle.com/s/2n5r35mriursy/tests/:testIntegration/bio.terra.service.filedata.DrsTest?top-execution=1))

Original attempt (wait before checking, then fail if inaccessible):
-  **1h 4m** with flaky and failing tests ([Gradle](https://scans.gradle.com/s/2n5r35mriursy/tests/:testIntegration/bio.terra.service.filedata.DrsTest?top-execution=1))

Wait before checking in test setup but use snapshot preview to fetch DRS ID:
- **39m** with flaky tests ([Gradle](https://scans.gradle.com/s/6znqwvp3ebnmg/tests/:testIntegration/bio.terra.service.filedata.DrsTest?top-execution=1))

Disable check in test setup and use snapshot preview to fetch DRS ID:
- **11m** with no flaky or failing tests ([Gradle](https://scans.gradle.com/s/27luu5xm6o6fw/tests/:testIntegration/bio.terra.service.filedata.DrsTest?top-execution=1))

okotsopoulos committed:
[DR-2925] Wait before checking for BQ access in certain tests

Matching Doug's recommendation, we now wait 3 minutes before checking for BQ access in affected test assertions to avoid poisoning Google's cache.
We've seen flakiness and failures in DrsTest runs specifically, so only adding in these waits to assertions used in those tests for now.

okotsopoulos committed:
If cloud resource inaccessible after initial wait, fail tests

While here, refactored BigQueryFixtures.hasAccess to void method assertBqDatasetAccessible:
this method performs its own assertions, so callers do not need a boolean representation of accessibility to reassert.

nmalfroy reviewed (approved):
looks great! (unfortunate we need it but here we are :) )

okotsopoulos committed:
Tests should use snapshot preview to fetch DRS IDs

We previously would query BigQuery directly to obtain a DRS ID in a snapshot, but unreliable IAM propagation has left this behavior flaky even with an initial 3 minute wait.
We should see more reliable behavior by obtaining a snapshot's DRS IDs via the snapshot preview endpoint: TDR proxies the request to BigQuery in this case.

- Replaced BigQueryFixtures.queryForDrsId -> DataRepoFixtures.retrieveDrsIdFromSnapshotPreview
- Previously these calls needed to be run as the snapshot custodian, but we can run them as the snapshot reader now
- Break up the remaining 3 minute sleep to log every 30 seconds (easier to follow the behavior in the logs this way)

okotsopoulos commented on src/test/java/bio/terra/integration/BigQueryFixtures.java:
```diff
@@ -142,51 +132,43 @@ public static String makeTableRef(DatasetModel datasetModel, String tableName) {
+  public static void assertBqDatasetAccessible(
+      BigQuery bigQuery, String dataProject, String bqDatasetName) throws Exception {
+    int sleptSeconds = 0;
+    while (sleptSeconds < WAIT_FOR_ACCESS_SECONDS) {
```
A concrete example of the logs emitted:

```
    2023-02-07 14:57:47,229 INFO  [Test worker] b.t.s.f.g.firestore.EncodeFixture: Checking BQ access for snapshot encodefiletest_snapshotdcc64158_4f35_459d_b41f_658f0c75919b in data project datarepo-dev-9b19ca38 with BQ dataset named encodefiletest_snapshotdcc64158_4f35_459d_b41f_658f0c75919b
    2023-02-07 14:57:47,667 INFO  [Test worker] b.terra.integration.BigQueryFixtures: Slept 0 seconds prior to checking for BigQuery access at 180 seconds
    2023-02-07 14:58:17,672 INFO  [Test worker] b.terra.integration.BigQueryFixtures: Slept 30 seconds prior to checking for BigQuery access at 180 seconds
    2023-02-07 14:58:47,675 INFO  [Test worker] b.terra.integration.BigQueryFixtures: Slept 60 seconds prior to checking for BigQuery access at 180 seconds
    2023-02-07 14:59:17,678 INFO  [Test worker] b.terra.integration.BigQueryFixtures: Slept 90 seconds prior to checking for BigQuery access at 180 seconds
    2023-02-07 14:59:47,682 INFO  [Test worker] b.terra.integration.BigQueryFixtures: Slept 120 seconds prior to checking for BigQuery access at 180 seconds
    2023-02-07 15:00:17,688 INFO  [Test worker] b.terra.integration.BigQueryFixtures: Slept 150 seconds prior to checking for BigQuery access at 180 seconds
    2023-02-07 15:00:47,697 INFO  [Test worker] b.terra.integration.BigQueryFixtures: Slept 180 seconds: checking for BigQuery access now
```

snf2ye reviewed (approved):
LGTM - I wonder if we could look at removing some of the assertBqDatasetAccessible() checks.

nmalfroy commented on src/test/java/bio/terra/integration/DataRepoFixtures.java:
```diff
@@ -687,6 +691,50 @@ public SnapshotModel getSnapshot(
+    SnapshotPreviewModel validated =
+        validateResponse(response, "snapshot preview for DRS ID", HttpStatus.OK, null);
+
+    assertThat("Got one row", validated.getResult(), hasSize(1));
```
Do we want to enforce this or just return the first one we find?

okotsopoulos replied:
So I was trying to port over the assertions and general functionality from the old BQ fixture, and this was something we asserted.  I think it's reasonable to include in this new version: it would be unexpected to have an empty result set at this stage, and we've already limited the results to 1.

nmalfroy commented on src/test/java/bio/terra/integration/DataRepoFixtures.java:
```diff
@@ -687,6 +691,50 @@ public SnapshotModel getSnapshot(
+    String drsUri = (String) ((LinkedHashMap) validated.getResult().get(0)).get(column);
+    assertThat("DRS URI was found", drsUri, notNullValue());
+
+    return DrsService.getLastNameFromPath(drsUri);
```
Is this to specific to the test to be in the fixture?  (e.g. feels like we should just return the drs id (if anything, we could validate that the string is a valid drs id).  If we return a drs id, we could actually return it as a `DrsId` object

nmalfroy reviewed (changes requested):
Looks great!  Just a couple of comments about the fixture method I think being too closely tied to the particular test that's using it

okotsopoulos commented:
> LGTM - I wonder if we could look at removing some of the assertBqDatasetAccessible() checks.

@snf2ye , I'm glad you brought this up!

<img width="700" alt="Screen Shot 2023-02-07 at 3 35 30 PM" src="https://user-images.githubusercontent.com/79769153/217359664-57a8c794-a709-4e8f-9124-aa77ec5004b7.png">

I was able to get the DRS tests to pass locally, but they were flaky.  3/6 attempts failed on the test setup accessibility check, and the accessibility check adds 3 minutes to each attempt.

- If we removed the accessibility check, overall test runtime would have gone from 38m 16s -> 10m 35s (none of the failed attempts and shorter test setup in all runs).
- Even if we don't disable this assertion, upgrading to JUnit 5 and being able to share setup across a test class would also save us significant runtime… setup right now takes about 6 minutes.

Given this, **I am inclined to disable the assertBqDatasetAccessible() checks used in DRS test setup**.  I think that the flakiness, known IAM propagation delays, and long test runtimes all outweigh their potential value at this time.  What do you think, Shelby / @nmalfroy ?

nmalfroy reviewed (approved):
Caught up off line and I think the current fixture approach makes sense

okotsopoulos commented:
After chatting with @nmalfroy (thanks!) we decided to go ahead and disable the accessibility check used in DRS test set up: we know that IAM propagation is unreliable and will benefit from the increased stability in our tests.  As a point of comparison, the DRS test suite ran locally in 10m 56s with no failed attempts when we disabled the check.

<img width="707" alt="Screen Shot 2023-02-07 at 4 46 27 PM" src="https://user-images.githubusercontent.com/79769153/217374399-f90b74b5-9b09-42e3-8dda-8506656380ca.png">

I did leave the underlying code in place in case we end up wanting to enable it again.

okotsopoulos committed:
DRS test set-up disables BigQuery dataset accessibility check

We were still seeing significant test flakiness even with the 3 minute wait, on top of long test runtimes.
We decided that IAM propagation is unpredictable enough as to warrant disabling this check in automated tests.
Downstream test code does not interact directly with BigQuery: an earlier commit in this PR obtains a snapshot's DRS ID via TDR's snapshot preview endpoint.
